### ISSUE TITLE
Segmentation fault when using 4 GPUs for training

### ISSUE THREAD

**Issue by thumbe3** [facebookresearch/dlrm#42]
Specs:

```
Python version: 3.6.8
Pytorch version: 1.4.0
4 v100 GPUs
Cuda version: 10.1
Nvidia Driver Version: 418.87.00 

```
I added thefollowing line in dlrm_s_pytorch.py

`import faulthandler; faulthandler.enable()`

and used the following command to run the code

`python3 -X faulthandler dlrm_s_pytorch.py --arch-sparse-feature-size=16 --arch-mlp-bot="13-512-256-64-16" --arch-mlp-top="512-256-1" --data-generation=dataset --data-set=kaggle --raw-data-file=/path-to-data --processed-data-file=/path-to-npz-file  --loss-function=bce --round-targets=True --learning-rate=0.1 --mini-batch-size=64  --test-freq 0 --print-freq=1024 --print-time --use-gpu`

It executes for some iterations(variable: different during different runs)
and then fails with a segmentation fault.
Here is a sample output

```
Using 4 GPU(s)...
Reading pre-processed data=/users/ushmal/kaggleAdDisplayChallenge_processed.npz
Sparse features= 26, Dense features= 13
Reading pre-processed data=/users/ushmal/kaggleAdDisplayChallenge_processed.npz
Sparse features= 26, Dense features= 13
time/loss/accuracy (if enabled):
Finished training it 1024/613937 of epoch 0, 51.35 ms/it, loss 0.520202, accuracy 75.478 %
Finished training it 2048/613937 of epoch 0, 28.98 ms/it, loss 0.506464, accuracy 76.196 %
Finished training it 3072/613937 of epoch 0, 29.48 ms/it, loss 0.505029, accuracy 76.314 %
Finished training it 4096/613937 of epoch 0, 30.34 ms/it, loss 0.494111, accuracy 76.935 %
Finished training it 5120/613937 of epoch 0, 30.36 ms/it, loss 0.496054, accuracy 76.781 %
Finished training it 6144/613937 of epoch 0, 30.44 ms/it, loss 0.487835, accuracy 77.235 %
Finished training it 7168/613937 of epoch 0, 30.65 ms/it, loss 0.486214, accuracy 77.292 %
Fatal Python error: Segmentation fault

Thread 0x00007f64c1a25700 (most recent call first):

Thread 0x00007f64c2226700 (most recent call first):

Current thread 0x00007f64c2a27700 (most recent call first):

Thread 0x00007f64c3a29700 (most recent call first):
  File "/users/ushmal/.local/lib/python3.6/site-packages/torch/cuda/comm.py", line 165 in gather
  File "/users/ushmal/.local/lib/python3.6/site-packages/torch/nn/parallel/_functions.py", line 68 in forward
  File "/users/ushmal/.local/lib/python3.6/site-packages/torch/nn/parallel/_functions.py", line 101 in backward
  File "/users/ushmal/.local/lib/python3.6/site-packages/torch/autograd/function.py", line 77 in apply

Thread 0x00007f64c3228700 (most recent call first):

Thread 0x00007f65f70b2740 (most recent call first):
  File "/users/ushmal/.local/lib/python3.6/site-packages/torch/autograd/__init__.py", line 99 in backward
  File "/users/ushmal/.local/lib/python3.6/site-packages/torch/tensor.py", line 195 in backward
  File "dlrm_s_pytorch.py", line 814 in <module>
Segmentation fault (core dumped)

```

When using 2 GPUs or a single GPU  the segmentation fault doesn't arise even after 100000 iterations.
Thanks!

**Comment by mnaumovfb** [comment]
Thank you for reporting this issue. We will take a look at it. 

I wanted to clarify a couple of things:
1. Do you see the crash without the faulthandler (i.e. no new line and no command line prefix)?
2. I wanted to confirm that you see no crash on a single or two GPUs as mentioned above.

**Comment by thumbe3** [comment]
Thanks for the response

1) Yes, I see the crash without the fault handler as well
2) Yes, I don't see a crash wheb using single as well as two gpus

**Comment by mnaumovfb** [comment]
We are still investigating what is happening with the newer version of the framework. However, I wanted to mention that I ran the latest version of the code without a crash, with the following older version of PyTorch and CUDA:
>>> import torch 
>>> print(torch.__version__)
1.1.0
>>> print(torch.version.cuda)
9.0.176

**Comment by albanD** [comment]
This might be caused by this issue in pytorch: https://github.com/pytorch/pytorch/issues/31906 
Could you try installing the nightly build and check that this is fixed please?

**Comment by mnaumovfb** [comment]
I believe that the above PyTorch issue was indeed the source of this bug. The back trace from the core file generated during the crash is very similar to the one reported above. 

I have rerun both Criteo Kaggle and Terabyte datasets on 8 GPUs with the nightly PyTorch build, where the above issue is fixed, and I do not see the crash any more. 

Please give it a try and let me know if it is resolved for you as well. You can obtain the nightly PyTorch build using for example the following command with the conda environment
"conda install pytorch torchvision cudatoolkit=10.1 -c pytorch-nightly"

**Comment by thumbe3** [comment]
Thanks for solving the issue.!I currently don't have the resources to rerun the experiment. I am closing the issu since it seems like the issue is fixed.

**Comment by Adamits** [comment]
Is this fix in the pytorch 1.4.0 pypy distribution? I am getting pretty much the same backtrace when running `backward()` on 4 GPUs with `DataParallel`. Is there a build that I should be using? I am having an unrelated issue training my model with pytorch 1.5.0

**Comment by mnaumovfb** [comment]
I don't think the fix made it into the PyTorch 1.4.0. Can you please try using 1.5 or the nightly version? Let me know if it works for you.

**Comment by Adamits** [comment]
Hi,

Ok, I was trying 1.4.0 because of https://github.com/huggingface/transformers/issues/3936. My experiment is using HuggingFace. I will try the nightly version (it errored the last time I tried though.)

**Comment by Davidrjx** [comment]
i met with same issue for setting below,
```
PyTorch version: 1.4.0
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Ubuntu 18.04.2 LTS
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
CMake version: version 3.14.5

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration:
GPU 0: GeForce RTX 2080 Ti
GPU 1: GeForce RTX 2080 Ti
GPU 2: GeForce RTX 2080 Ti
GPU 3: GeForce RTX 2080 Ti

Nvidia driver version: 418.43
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.0

Versions of relevant libraries:
[pip3] numpy==1.17.1
[pip3] torch==1.4.0
[pip3] torchvision==0.5.0
[conda] Could not collect
```
any result about nightly version verification above?

**Comment by mnaumovfb** [comment]
As we were discussing above, can you please try using 1.5 or the nightly version of PyTorch?

**Comment by Davidrjx** [comment]
> As we were discussing above, can you please try using 1.5 or the nightly version of PyTorch?
> 
> I'm assuming it has worked in the past because there was no follow up, but let me know if it works for you.

very good, thank your reply @mnaumovfb, and i will make some work about your suggestion with my collegue.
